add Identifier to InvalidSignatureShare and split Error type

conradoplg opened the pull request:
This is an alternative to #182 which is cleaner but more invasive. I had to split the Error enum into Error, FieldError and GroupError.

I added the `Debug` bound to the ciphersuite because assertions need it. Alternatively we can use `debugless_unwrap()` which we already use elsewhere, and remove the bound.

dconnolly commented on frost-core/src/error.rs:
```diff
@@ -53,7 +40,10 @@ pub enum Error {
     /// Signature share verification failed.
     #[error("Invalid signature share.")]
-    InvalidSignatureShare,
+    InvalidSignatureShare {
+        /// The identifier of the signer whose share validation failed.
+        signer: Identifier<C>,
+    },
```
👍

dconnolly commented on frost-core/src/error.rs:
```diff
@@ -72,4 +62,37 @@ pub enum Error {
+    GroupError(#[from] GroupError),
+}
+
+/// An error related to a scalar Field.
+#[non_exhaustive]
+#[derive(Error, Debug, Copy, Clone, Eq, PartialEq)]
+pub enum FieldError {
+    /// The encoding of a group scalar was malformed.
+    #[error("Malformed scalar encoding.")]
+    MalformedScalar,
+    /// This scalar MUST NOT be zero.
+    #[error("Invalid for this scalar to be zero.")]
+    InvalidZeroScalar,
+}
```
👍

dconnolly commented on frost-core/src/error.rs:
```diff
@@ -72,4 +62,37 @@ pub enum Error {
+    InvalidZeroScalar,
+}
+
+/// An error related to an elliptic curve Group.
```
```suggestion
/// An error related to a Group or one of its Elements.
```

dconnolly commented on frost-core/src/frost.rs:
```diff
@@ -44,8 +44,10 @@ where
+    ) -> Result<Self, Error<C>> {
+        <<C::Group as Group>::Field>::deserialize(&bytes)
+            .map(|scalar| Self(scalar))
+            .map_err(|e| e.into())
```
Is this needed because the original `e` may be `FieldError` or `GroupError`?

conradoplg replied:
Exactly, in this case it's always `FieldError`, but we want an `Error`. Since `Error` has a `FieldError(#[from] FieldError),` it has a `From` implementation that allows us to just call `.into()`.

dconnolly commented on frost-core/src/frost/identifier.rs:
```diff
@@ -109,11 +109,11 @@ impl<C> TryFrom<u16> for Identifier<C>
     fn try_from(n: u16) -> Result<Identifier<C>, Self::Error> {
         if n == 0 {
-            Err(Self::Error::InvalidZeroScalar)
+            Err(FieldError::InvalidZeroScalar.into())
```
👍

dconnolly commented on frost-core/src/frost/keys.rs:
```diff
@@ -1,4 +1,5 @@
 //! FROST keys, keygen, key shares
+#![allow(clippy::type_complexity)]
```
😅

dconnolly commented on frost-core/src/frost/round2.rs:
```diff
@@ -86,11 +88,13 @@ where
             != (group_commitment_share.0 + (public_key.0 * challenge.0 * lambda_i))
         {
-            return Err(Error::InvalidSignatureShare);
+            return Err(Error::InvalidSignatureShare {
+                signer: self.identifier,
+            });
```
🎉

dconnolly reviewed (approved):
Looks great! 😎

dconnolly reviewed (changes requested):
Looks great, but may have missed some `C`s 😅

<img width="449" alt="image" src="https://user-images.githubusercontent.com/552961/204971407-834755ea-292e-41f7-9ec1-f1c2e9dfe4b8.png">

conradoplg commented:
> Looks great, but may have missed some `C`s sweat_smile
> 
> <img alt="image" width="449" src="https://user-images.githubusercontent.com/552961/204971407-834755ea-292e-41f7-9ec1-f1c2e9dfe4b8.png">

Oops, that's because `secp256k1` was merged after I created this PR. I've just updated it.

conradoplg commented:
@dconnolly I added the Identifier serialization methods here too, and updated it to work with the recently merged Ed448. This should be good to go, and we can close #182

dconnolly commented on frost-core/src/frost/identifier.rs:
```diff
@@ -19,10 +19,23 @@ impl<C> Identifier<C>
+    pub fn deserialize(
+        buf: &<<C::Group as Group>::Field as Field>::Serialization,
+    ) -> Result<Self, Error<C>> {
+        let scalar = <<C::Group as Group>::Field>::deserialize(buf)?;
```
```suggestion

        let scalar = <<C::Group as Group>::Field>::deserialize(buf)?;
```

dconnolly replied:
lol, this is against rustfmt

<img width="542" alt="image" src="https://user-images.githubusercontent.com/552961/207157766-453ef971-d8b0-47ff-98ea-2584475960aa.png">

dconnolly commented on frost-core/src/frost/identifier.rs:
```diff
@@ -19,10 +19,24 @@ impl<C> Identifier<C>
+    pub fn deserialize(
+        buf: &<<C::Group as Group>::Field as Field>::Serialization,
+    ) -> Result<Self, Error<C>> {
+
+        let scalar = <<C::Group as Group>::Field>::deserialize(buf)?;
+        if scalar == <<C::Group as Group>::Field>::zero() {
```
```suggestion
        let scalar = <<C::Group as Group>::Field>::deserialize(buf)?;
        
        if scalar == <<C::Group as Group>::Field>::zero() {
```

dconnolly commented on frost-core/src/frost/identifier.rs:
```diff
@@ -19,10 +19,24 @@ impl<C> Identifier<C>
+    pub fn deserialize(
+        buf: &<<C::Group as Group>::Field as Field>::Serialization,
+    ) -> Result<Self, Error<C>> {
+        let scalar = <<C::Group as Group>::Field>::deserialize(buf)?;
+        
+        if scalar == <<C::Group as Group>::Field>::zero() {
```
```suggestion
        let scalar = <<C::Group as Group>::Field>::deserialize(buf)?;
        if scalar == <<C::Group as Group>::Field>::zero() {
```
